dll/win32/wdmaud.drv/mmixer.c

```diff
@@ -805,6 +805,123 @@ WdmAudGetWavePositionByMMixer(
     return MMSYSERR_NOTSUPPORTED;
 }
 
+MMRESULT
+WdmAudGetVolumeByMMixer(
+    _In_ struct _SOUND_DEVICE_INSTANCE* SoundDeviceInstance,
+    _In_ DWORD DeviceId,
+    _Out_ PDWORD pdwVolume)
+{
+    MMRESULT Result;
+    MIXERLINE MixLine;
+    MIXERCONTROL MixControl;
+    MIXERLINECONTROLS MixLineControls;
+    MIXERCONTROLDETAILS MixControlDetails;
+    MIXERCONTROLDETAILS_UNSIGNED MixControlDetailsU[2];
+
+    MixLine.cbStruct = sizeof(MIXERLINE);
+    MixLine.dwComponentType = MIXERLINE_COMPONENTTYPE_DST_SPEAKERS;
+
+    /* Get line info */
+    Result = WdmAudGetLineInfo(SoundDeviceInstance->Handle,
+                               DeviceId,
+                               &MixLine,
+                               MIXER_OBJECTF_MIXER | MIXER_GETLINEINFOF_COMPONENTTYPE);
+    if (!MMSUCCESS(Result))
+        return TranslateInternalMmResult(Result);
+
+    MixLineControls.cbStruct = sizeof(MIXERLINECONTROLS);
+    MixLineControls.dwLineID = MixLine.dwLineID;
+    MixLineControls.dwControlType = MIXERCONTROL_CONTROLTYPE_VOLUME;
+    MixLineControls.cControls = 1;
+    MixLineControls.cbmxctrl = sizeof(MIXERCONTROL);
+    MixLineControls.pamxctrl = &MixControl;
+
+    /* Get line controls */
+    Result = WdmAudGetLineControls(SoundDeviceInstance->Handle,
+                                   DeviceId,
+                                   &MixLineControls,
+                                   MIXER_OBJECTF_MIXER | MIXER_GETLINECONTROLSF_ONEBYTYPE);
+    if (!MMSUCCESS(Result))
+        return TranslateInternalMmResult(Result);
+
+    MixControlDetails.cbStruct = sizeof(MIXERCONTROLDETAILS);
+    MixControlDetails.dwControlID = MixControl.dwControlID;
+    MixControlDetails.cChannels = MixLine.cChannels;
+    MixControlDetails.cMultipleItems = 0;
+    MixControlDetails.cbDetails = sizeof(MIXERCONTROLDETAILS_UNSIGNED);
+    MixControlDetails.paDetails = MixControlDetailsU;
```

Comment:
According to MSDN
https://learn.microsoft.com/en-us/windows/win32/api/mmeapi/ns-mmeapi-mixercontroldetails
> For controls other than MIXERCONTROL_CONTROLF_MULTIPLE types, the size of this buffer is the product of the cChannels and cbDetails members of the MIXERCONTROLDETAILS structure. For controls other than MIXERCONTROL_CONTROLF_MULTIPLE types, the size of this buffer is the product of the cChannels and cbDetails members of the MIXERCONTROLDETAILS structure.

Here it seems that your array has a fixed size (two elements), but how does this relate to the `cChannels` value (which you set to be equal to `MixLine.cChannels`) above?

Reply:
@HBelusca according to `MIXERLINE` structure documentation:
https://learn.microsoft.com/en-us/windows/win32/api/mmeapi/ns-mmeapi-mixerline
>Maximum number of separate channels that can be manipulated independently for the audio line. The minimum value for this field is 1 because a line must have at least one channel.
Most modern audio cards for personal computers are stereo devices; for them, the value of this member is 2.
Channel 1 is assumed to be the left channel; channel 2 is assumed to be the right channel.
A multichannel line might have one or more uniform controls (controls that affect all channels of a line uniformly) associated with it.

`cChannels` member of this structure contains the actual number of channels, and it is supposed to have the same purpose and same range of values (1, 2 or more) as `cChannels` member of `MIXERCONTROLDETAILS`. So do we actually need to always use a hardcoded value, even when the actual number of channels is different?

Reply:
The hardcoded value 1 is used only for `MIXERCONTROL_CONTROLTYPE_MUTE` or `MIXERCONTROL_CONTROLTYPE_MUX`, but not for `MIXERCONTROL_CONTROLTYPE_VOLUME`.
Just look at our MMSys code. Here's `cChannels` value is taken from `MIXERLINE`:
https://git.reactos.org/?p=reactos.git;a=blob;f=dll/cpl/mmsys/volume.c;hb=f69e256376e939e63029c8d8be8b97eaa206eec0#l134
And then it is stored in `MIXERCONTROLDETAILS.cChannels` member:
https://git.reactos.org/?p=reactos.git;a=blob;f=dll/cpl/mmsys/volume.c;hb=f69e256376e939e63029c8d8be8b97eaa206eec0#l184
same as in other functions.
It is only hardcoded here:
https://git.reactos.org/?p=reactos.git;a=blob;f=dll/cpl/mmsys/volume.c;hb=f69e256376e939e63029c8d8be8b97eaa206eec0#l83
but that is actually Mute Control, not Volume.
Similarly is in SndVol32, so why not do the same here?

Reply:
According to the docs you linked,
![Снимок экрана от 2024-05-23 10-42-06](https://github.com/reactos/reactos/assets/26385117/100971c8-b6ab-41fb-9f3c-b229645f0980)
it can take a value from `MIXERLINE.cChannels`. This is only a valid for our case, because we are not using `MIXEROCNTROL_CONTROLTYPE_CUSTOM` or `MIXERCONTROL_CONTROLTYPE_UNIFORM`. We are using `MIXERCONTROL_CONTROLTYPE_VOLUME` here, as you can see in the code above.

Reply:
The only a thing I probably need to do, is multiply the value of `cbDetails` to `cChannels`, but then it will fail in MMixer, since that member is validated there.

Reply:
> So do we actually need to always use a hardcoded value, even when the actual number of channels is different?

That's what I'm not fully sure indeed. The hardcoded two-element in the array looks strange to me, but I'm not fully versed in the multimedia/sound stuff to have a definite answer.

Reply:
As you can see, I'm using two elements of the array for getting the values from/storing the values in two channels appropriately, left and right. Single structure has only one `dwValue` member, so I can't store both of them in one member or simply assign input `dwVolume` to/get output `pdwVolume` from it. I did not try to do it yet, but I'm not sure it is correct, since MSDN states the following:
> The low-order word of this location contains the left-channel volume setting, and the high-order word contains the right-channel setting.

See
https://learn.microsoft.com/en-us/windows/win32/api/mmeapi/nf-mmeapi-waveoutgetvolume
and
https://learn.microsoft.com/en-us/windows/win32/api/mmeapi/nf-mmeapi-waveoutsetvolume
for Get and Set functions appropriately.

Reply:
In any case, I wonder what you see as the value of `MixLine.cChannels` when you run your code in ReactOS (or in Windows) with the debugger.

Reply:
OK, just debugged it. Its value always equals to `2`, as it's an actual number of channels (at least when calling `SetVolume` function):
`(dll/win32/wdmaud.drv/mmixer.c:903) MixLine.cChannels 2
(dll/win32/wdmaud.drv/mmixer.c:903) MixLine.cChannels 2
(dll/win32/wdmaud.drv/mmixer.c:903) MixLine.cChannels 2
(dll/win32/wdmaud.drv/mmixer.c:903) MixLine.cChannels 2
(dll/win32/wdmaud.drv/mmixer.c:903) MixLine.cChannels 2`
I'm sure it's correct value in this case.
Seems I should multiply `cbDetails` member on this `cChannels` if the value is greater than `1`, according to MSDN documentation, but then I need to update the check in the appropriate MMIXER routine too, otherwise it will fail.

Reply:
> Seems I should multiply cbDetails member on this cChannels if the value is greater than 1, according to MSDN documentation, but then I need to update the check in the appropriate MMIXER routine too, otherwise it will fail.

Seems in case I'll do it, I also should update all other occurrences of the similar code in another components (in a separate PR), to avoid failing that.

Reply:
> I wonder what would happen on a multi-stereo system on real hardware though

Unfortunately I currently have no hardware with compatible (non-HD audio) card. :slightly_frowning_face: MMixer mode isn't testable with HD audio, so it needs Legacy one to be implemented. I would test it if I'd have such a hardware, but why you are sure it can have a different result in ROS/XP/2003 on real hardware? :thinking:

Reply:
I would suggest you ask on mattermost or discord. Maybe @DarkFire01 has something to test, who knows? :thinking: :joy:

Reply:
> I would suggest you ask on mattermost or discord. Maybe @DarkFire01 has something to test, who knows? 🤔 😂

OK, will do.

Reply:
> > I would suggest you ask on mattermost or discord. Maybe @DarkFire01 has something to test, who knows? 🤔 😂
> 
> OK, will do.

Done: https://chat.reactos.org/reactos/pl/1icf1z1rwpfu9pc7sk9ehbjqth. :wink: